relayer/src/supervisor/spawn.rs

```diff
@@ -514,22 +508,34 @@ impl<'a> SpawnContext<'a> {
                 )
                 .then(|| debug!("spawned Client worker: {}", client_object.short_name()));
 
-            // create the Packet object and spawn worker
-            let path_object = Object::Packet(Packet {
-                dst_chain_id: counterparty_chain.id(),
-                src_chain_id: chain.id(),
-                src_channel_id: channel.channel_id,
-                src_port_id: channel.port_id,
-            });
+            let outstanding_packets =
+                !unreceived_packets(chain.as_ref(), counterparty_chain.as_ref(), channel.clone())?
+                    .is_empty()
+                    || !unreceived_acknowledgements(
+                        chain.as_ref(),
+                        counterparty_chain.as_ref(),
+                        channel.clone(),
+                    )?
+                    .is_empty();
+
+            if outstanding_packets {
+                // create the Packet object and spawn worker
+                let path_object = Object::Packet(Packet {
+                    dst_chain_id: counterparty_chain.id(),
+                    src_chain_id: chain.id(),
+                    src_channel_id: channel.channel_id,
+                    src_port_id: channel.port_id,
+                });
```

Comment:
The `Packet` object we're creating (line 523) is for the relaying path from `chain` (source chain) to `counterparty_chain` (destination chain). The worker for this object will be responsible for relaying packets based on events generated by source  `chain` and having destination `counterparty_chain`.

This worker will handle packets flowing strictly for the direction `chain` -> `counterparty_chain`.

The problem is that the logic in `outstanding_packets` algorithm (line 511-519) is not consistent with the direction of the worker. More specifically, the direction computed by `outstanding_packets` is reversed. When we compute `outstanding_packets`, we are considering the two following cases:

1. `unreceived_packets` for the direction `counterparty_chain` -> `chain`
2. `unreceived_acknowledgements` for the direction `counterparty_chain` -> `chain`.

To fix the problem, we need to check for unreceived packets and unreceived acknowledgements in the opposite direction, for packets on path `chain` -> `counterparty_chain`. I am leaving a suggestion for a patch below:

```suggestion
            let outstanding_packets =
                !unreceived_packets(counterparty_chain.as_ref(), chain.as_ref(), channel.clone())?
                    .is_empty()
                    || !unreceived_acknowledgements(
                        counterparty_chain.as_ref(),
                        chain.as_ref(),
                        channel.clone(),
                    )?
                    .is_empty();

            if outstanding_packets {
                // create the Packet object and spawn worker
                let path_object = Object::Packet(Packet {
                    dst_chain_id: counterparty_chain.id(),
                    src_chain_id: chain.id(),
                    src_channel_id: channel.channel_id,
                    src_port_id: channel.port_id,
                });
```

Reply:
Thanks for noticing and helping fix this bug @adi! That's a good catch! 👌 
Sorry for not testing this well enough and relying only on the E2E tests. Do you think we should create an issue to add a test for this in `run.py`?

Reply:
I don't actually understand why the E2E tests did not catch this. It's possible that it's because there were packets pending on _both_ directions of a path, and then workers were spawned for both directions, which masked the problem.

> Do you think we should create an issue to add a test for this in run.py?

Your suggestion sounds very good! But it would be good to confirm what is the reason why run.py was unable to catch this corner-case in the first place, then we can translate this into an issue.

Reply:
I just tested locally and I can confirm that the original bug has been fixed. 👍 Still trying to figure out why the E2E tests didn't catch it.

Reply:
So the E2E tests do fail one commit earlier even on multiple retries (at least locally) -> 
```python
Traceback (most recent call last):
  File "/relayer/e2e/run.py", line 276, in <module>
    main()
  File "/relayer/e2e/run.py", line 247, in main
    passive_packets(config, ibc0, ibc1, port_id, ibc0_chan_id, ibc1_chan_id)
  File "/relayer/e2e/run.py", line 98, in passive_packets
    assert (len(unreceived) == 0), (unreceived,
AssertionError: ([5, 6, 7], 'unreceived acks mismatch (expected 0)')
```

I think this happened before as well, but we ignored it thinking it was caused by #1216 (that was fixed recently). Still investigating the logs...

Reply:
I did some more testing and found that the `unreceived_packets()` function expects the channel on the party's side as it's 3rd arg ->
```rust
pub fn unreceived_packets(
    chain: &dyn ChainHandle,
    counterparty_chain: &dyn ChainHandle,
    channel: IdentifiedChannelEnd,
) -> Result<Vec<u64>, Error>;
```
But when we're calling it from `spawn_workers_for_channel()` ->
```rust
unreceived_packets(counterparty_chain.as_ref(), chain.as_ref(), channel.clone())?;
```
We're giving it the channel on the other chain. So we should've instead given it the `counterparty_channel` as the 3rd arg. Just fixed it.